### PR TITLE
Added (Ord a,Ord b) => Ord (Gr a b) instances to the Tree.hs and Patrici...

### PR DESCRIPTION
**Pull request by trevorcook**
...aTree.hs definitions of Gr. 

I'm playing with labeling nodes with graphs, and need to make them members of Ord so that I can use NodeMap. Could have used a generic "deriving(Ord)" clause but i don't know how that plays with the ghc pragma.

### CODE REVIEW

**Comment by ivan-m** [comment]
I don't think an `Ord` instance for graphs makes sense, as the ordering of graphs is a partially ordered set (no pun intended!).

(And even if it was to be implemented solely as a "it's useful as a Haskell function to be able to do this for use with things that require an Ord instance, it would make more sense to implement it using `compare` rather than `<=`.)

**Comment by trevorcook** [comment]
Ok, thanks. 

Yes, I'm using a comparison for the sake of putting graphs in a map. I suppose you don't want to have a user mistakenly treat Ord (Gr a b) as mathematically relevant. This is good, it might be misleading. I have to say, I don't see much room for harm stemming from this abuse. But hey, people will use your code in unexpected ways, right? Regardless, this Ord implementation is flawed.

For the sake of discussion, I'd also like to ask if the case for an Ord instance would be more compelling with an associated Enum instance? Especially, instance Enum (Gr () ()). This is something else I would have found useful in my work, but didn't implement because I didn't know how to deal with the graph homomorphisms.  

Something I think makes sense for an Enum is, notionally, [empty] ++ [one node] ++ [two nodes zero edges] ++ [two nodes one edge] ++ .. ++ concat [n nodes with homommrphisms of e edges | e<- 0..E]). The ordering/enumeration would be based: primarily on the number of nodes, secondarily on the number of edges, and finally on a dictionary style ordering on the size of the connected components of the graphs. As an example of that last rule, for n = 4 and e = 2 we have two graph homomorphisms, g1 and g2, with the (sorted) size (m for measure) of the connected components as, m g1 = [1,3], and m g2 = [2,2]. Dictionary wise [1,3] < [2,2], hence g1 < g2.

I realize I'm making a lot of arbitrary decisions above, a necessary step in going from a partial order to a total ordering, but maybe there's a case to be made that some decisions are at least as good as any other. What do you think? How about, implementation wise, some newtype wrappers which supply desired orderings a la Sum, Product, etc. wrappers for Monoids?

Sorry to talk your ears off, (write your eyes out?) Will you still be able to comment if I close the pull request? Thanks again.

**Comment by ivan-m** [comment]
I'm willing to be convinced that an Ord instance for graphs makes sense/might be worth adding (the definition you gave is inefficient though, rather than flawed per-se; does an Ord instance for Map make sense?); an Enum instance, however, just seems wrong.

So I don't think there's a need to close this pull request yet.

**Comment by trevorcook** [comment]
I think it might be flawed as well. The Eq instance sorts the adjacency lists of the nodes. If Ord doesn't do that, maybe we get 'compare g1 g2  -> LT' but 'g1==g2 -> True,' which is creepy.

An Ord instance for map? We can look at 'Map k v' as the set of partial functions from k to v right? I think domain theory would say that larger sets are more specified than smaller sets, so that seems like a rational starting point. And for maps of the same size? Again, arbitrary decision time. Perhaps based on a dictionary sort of (k,v) pairs. If not well founded then at least its common convention. I didn't check how containers implemented it.

I'm not convinced these Ord ideas are good. If we can't recognize them as definitely bad, ok, i lets keep the request open. Maybe there is a fgl-extras where this kind of thing should go? Well, that would mean that the constructor Gr would have to be exported, but that's pretty much against the grain of fgl as an abstract data type.

As for Enum, is it really so bad? It seems kind of tame to me compared to, for instance an enumeration of the rational numbers. Look at [Bird Tree](http://www.cs.ox.ac.uk/ralf.hinze/publications/Bird.pdf). A breadth first search of that tree will list off all the rational numbers. Its not the only way to do it, and that enumeration doesn't mesh with the ordering of rationals, but at least it hits them all. Likewise, we should hope to hit all graphs from size 0 to N, for any N. Seems feasible, conceptually at least. And we should be able to hope that the ordering of the graphs meshes with the enumeration in a way that enumeration and ordering of rationals doesn't.

I'll think on this some. Work with it in my stuff and let you know how it goes. Thanks for your consideration.

**Comment by ivan-m** [comment]
Right, yeah, that slipped my mind (I even forgot to do it for `Eq` at one stage).

**Comment by ivan-m** [comment]
Going by the mailing list thread, what about adding a newtype to `Data.Graph.Inductive.Graph` such that:

```.haskell
instance (Graph gr, Ord a, Ord b) => Ord (OrdGraph gr a b) where
  compare gr1 gr2 = ((compare `on` sort . labNodes) gr1 gr2) `mappend` ((compare `on` sort . labEdges) gr1 gr2)
```

**Comment by trevorcook** [comment]
So nice!

**Comment by ivan-m** [comment]
Realistically speaking, the call to sort for labNodes isn't needed.  It also requires checking that my recollection of the Monoid instance for Ordering is right (and make sure it's been in GHC for a while and not something new-ish).

(The `equal` function should be something like this, but the latest butchery I've added in for it is just to avoid adding the `Ord` constraints on `a` and `b` so as to avoid changing the type signature...)

**Comment by trevorcook** [comment]
Checking hackage, it is defined lexicographically since at least 4.0.0.0. In 4.8.0.0, the definition seems to have moved out of Data.Monoid, perhaps into GHC.Base? Anyways, still defined in 4.7.0.2 as:
instance Monoid Ordering where
        mempty         = EQ
        LT `mappend` _ = LT
        EQ `mappend` y = y
        GT `mappend` _ = GT

**Comment by cmaeder** [comment]
If you have a proper Ord instance the Eq instance should just be defined as `a == b = compare a b == EQ`. (P.S. Data.Set and Data.Map have proper (total!) Ord instances.)

**Comment by ivan-m** [comment]
@cmaeder you _could_ do that... but why?  Especially when it's possible there's a more efficient way of defining the Eq instance...

@trevorcook I meant define it in `Data.Graph.Inductive.Graph` and have it work for any instance of the `Graph` class rather than per-instance (also, the commit message is a bit of a mouthful!)

**Comment by cmaeder** [comment]
`a == b = compare a b == EQ` avoids code duplication and ensures consistency of Eq with Ord.

**Comment by trevorcook** [comment]
@ivan-m 
*commit message too verbose? haha, yeah.
*ghc told me Eq (gr a b) was needed. Your dare question the compiler! Seriously though, the reason is that the two specific Gr implementations (PatritiaTree.hs and Tree.hs) have (Eq a, Ord b) => Eq (Gr a b), but there is no rule for any Graph gr, i.e instance (Graph gr, Eq a, Ord b) => Eq (gr a b).
*Noted
*FlexibleContexts is needed by the inclusion of Eq (gr a b) in the context. This is considered a pretty harmless extension, right? I guess the way around would be to define the instance for the specific graph implementations.

@cmaeder Note that the class definition of Ord relies on Eq, so defining Eq in terms of Ord is, well I don't think it can be done.

**Comment by ivan-m** [comment]
Well, it wants the Eq constraint because Ord requires Eq, so you should add an explicit Eq instance (but please derive Read and Show).  You can do it via the Ord constraint as @cmaeder suggests (yes, it's valid) or directly (use `(==)` rather than `compare`).

This way you wouldn't need `FlexibleContexts` (yes, it's harmless... but it also isn't needed ;-)

A better commit message would be `Allow explicitly requiring Ord instance for graphs` with a longer description below about why it isn't directly added to the two instances (link to the mailing list thread, etc.).

**Comment by ivan-m** [comment]
Also, please add Haddock documentation to the `OrdGr` wrapper type.

**Comment by trevorcook** [comment]
Ivan-m, I sorry but I don't understand your explanation of how to get rid of `FlexibleContexts`. I'm wondering if you are misplacing `graph gr=>gr` for `Data.Inductive.PatriciaTree.Gr`? Otherwise, I must be really missing an important point. Here's my thinking. 

When I automatically derive `Eq` for `newtype OrdGr gr a b`, the compiler will create something like `instance Eq (gr a b) => Eq (OrdGr gr a b) where ...`, right? But, I don't know how to write my own instance which wouldn't rely on `Eq (gr a b)`, itself. And I don't know how to generally write an `instance (Graph gr, Ctxa a, Ctxb b) => Eq (gr a b)`. 

If I were to move the definition of `OrdGr` out of `Data.Graph.Inductive.Graph` to `Data.Graph.Inductive`, where the specific `Gr` is imported, then I could make the definitions based on `Gr a b` as opposed to `Graph gr => gr a b`, and things would work out fine. 

Otherwise, I don't see it. Please show me again.

**Comment by trevorcook** [comment]
Ok. That would have probably been easier for you to do all by your self. I appreciate you taking the time to explain it to me so that I could do it. Thanks!

**Comment by ivan-m** [comment]
No worries; this is how you learn, right? :)

Just do those last few cosmetic tweaks and then squash them all together into one big commit and we should be done!

**Comment by trevorcook** [comment]
I glubbed up the squash. Uggh. Any advice?

**Comment by ivan-m** [comment]
In your local repository, do something like `git rebase --interactive HEAD~5`, and then when that's done do `git push --force` to make sure the new versions of your commit are sent to GitHub.

**Comment by ivan-m** [comment]
Also, it looks like the long commit message is one big line; it's usually recommended that you line-wrap it at about [72 characters per line](http://git-scm.com/book/ch5-2.html).

**Comment by trevorcook** [comment]
Ok, I think that did it. For some reason using 'git rebase -i HEAD~N' wouldn't start where I thought it should be, using the hash instead of HEAD~N seems to be what I needed. Also the --force option. Thanks Again :)

**Comment by ivan-m** [comment]
One last nit-pick ;-)
